From: Leonardo Arena <rnalrd@alpinelinux.org>
Date: Tue, 22 Sep 2026 14:01:01 +0200
Subject: [PATCH] Avoid the deprecated libudev udev_queue API when checking
 whether a real udev daemon is active for udev_sync support

This removes the only use of the udev_queue_* API in the tree and
means libdevmapper.so no longer needs to link against libudev at all
for --enable-udev_sync support.

See: https://gitlab.alpinelinux.org/alpine/aports/-/work_items/18416
---

diff --git a/device_mapper/libdm-common.c b/device_mapper/libdm-common.c
index 4fb5b7f..24c1d79 100644
--- a/device_mapper/libdm-common.c
+++ b/device_mapper/libdm-common.c
@@ -31,7 +31,6 @@
 #  include <sys/types.h>
 #  include <sys/ipc.h>
 #  include <sys/sem.h>
-#  include <libudev.h>
 #endif

 #ifdef __linux__
@@ -49,6 +48,8 @@

 #define DEV_DIR "/dev/"

+#define UDEV_CONTROL_PATH "/run/udev/control"
+
 #ifdef UDEV_SYNC_SUPPORT
 #ifdef _SEM_SEMUN_UNDEFINED
 union semun
@@ -2323,30 +2324,13 @@ static int _check_semaphore_is_supported(void)

 static int _check_udev_is_running(void)
 {
-	struct udev *udev;
-	struct udev_queue *udev_queue;
-	int r;
-
-	if (!(udev = udev_new()))
-		goto_bad;
-
-	if (!(udev_queue = udev_queue_new(udev))) {
-		udev_unref(udev);
-		goto_bad;
-	}
+	int r = access(UDEV_CONTROL_PATH, F_OK) >= 0;

-	if (!(r = udev_queue_get_udev_is_active(udev_queue)))
+	if (!r)
 		log_debug_activation("Udev is not running. "
 				     "Not using udev synchronization code.");

-	udev_queue_unref(udev_queue);
-	udev_unref(udev);
-
 	return r;
-
-bad:
-	log_error("Could not get udev state. Assuming udev is not running.");
-	return 0;
 }

 static void _check_udev_sync_requirements_once(void)
diff --git a/lib/misc/lvm-wrappers.c b/lib/misc/lvm-wrappers.c
index 8f7570e..de28149 100644
--- a/lib/misc/lvm-wrappers.c
+++ b/lib/misc/lvm-wrappers.c
@@ -20,6 +20,8 @@
 #ifdef UDEV_SYNC_SUPPORT
 #include <libudev.h>

+#define UDEV_CONTROL_PATH "/run/udev/control"
+
 struct udev *_udev;

 int udev_init_library_context(void)
@@ -30,14 +32,11 @@ int udev_init_library_context(void)
 	if (getenv("DM_DISABLE_UDEV"))
 		return 0;

-	if (!(_udev = udev_new())) {
-		log_error("Failed to create udev library context.");
+	if (!udev_is_running())
 		return 0;
-	}

-	if (!udev_is_running()) {
-		udev_unref(_udev);
-		_udev = NULL;
+	if (!(_udev = udev_new())) {
+		log_error("Failed to create udev library context.");
 		return 0;
 	}

@@ -53,25 +52,9 @@ void udev_fin_library_context(void)

 int udev_is_running(void)
 {
-	struct udev_queue *udev_queue;
-	int r;
-
-	if (!_udev) {
-		log_debug_activation("Udev library context not set.");
-		goto bad;
-	}
-
-	if (!(udev_queue = udev_queue_new(_udev))) {
-		log_debug_activation("Could not get udev state.");
-		goto bad;
-	}
-
-	r = udev_queue_get_udev_is_active(udev_queue);
-	udev_queue_unref(udev_queue);
-
-	return r;
+	if (access(UDEV_CONTROL_PATH, F_OK) >= 0)
+		return 1;

-bad:
 	log_debug_activation("Assuming udev is not running.");
 	return 0;
 }
diff --git a/libdm/dm-tools/dmsetup.c b/libdm/dm-tools/dmsetup.c
index ec20ac9..eb925d6 100644
--- a/libdm/dm-tools/dmsetup.c
+++ b/libdm/dm-tools/dmsetup.c
@@ -44,7 +44,6 @@
 #  include <sys/types.h>
 #  include <sys/ipc.h>
 #  include <sys/sem.h>
-#  include <libudev.h>
 #endif

 /* FIXME Unused so far */
diff --git a/libdm/libdm-common.c b/libdm/libdm-common.c
index de82831..9ddb457 100644
--- a/libdm/libdm-common.c
+++ b/libdm/libdm-common.c
@@ -29,7 +29,6 @@
 #  include <sys/types.h>
 #  include <sys/ipc.h>
 #  include <sys/sem.h>
-#  include <libudev.h>
 #endif

 #ifdef __linux__
@@ -47,6 +46,8 @@

 #define DEV_DIR "/dev/"

+#define UDEV_CONTROL_PATH "/run/udev/control"
+
 #ifdef UDEV_SYNC_SUPPORT
 #ifdef _SEM_SEMUN_UNDEFINED
 union semun
@@ -2311,30 +2312,13 @@ static int _check_semaphore_is_supported(void)

 static int _check_udev_is_running(void)
 {
-	struct udev *udev;
-	struct udev_queue *udev_queue;
-	int r;
-
-	if (!(udev = udev_new()))
-		goto_bad;
-
-	if (!(udev_queue = udev_queue_new(udev))) {
-		udev_unref(udev);
-		goto_bad;
-	}
+	int r = access(UDEV_CONTROL_PATH, F_OK) >= 0;

-	if (!(r = udev_queue_get_udev_is_active(udev_queue)))
+	if (!r)
 		log_debug_activation("Udev is not running. "
 				     "Not using udev synchronization code.");

-	udev_queue_unref(udev_queue);
-	udev_unref(udev);
-
 	return r;
-
-bad:
-	log_error("Could not get udev state. Assuming udev is not running.");
-	return 0;
 }

 static void _check_udev_sync_requirements_once(void)
